File: Leetcode 1801-1900/1807/1807.py
# Leetcode 1807: Evaluate the Bracket Pairs of a String
# https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string/
# Solved on 27th of November, 2025
class Solution:
    def evaluate(self, s: str, knowledge: list[list[str]]) -> str:
        """
        Evaluates a string by replacing bracketed keys with their corresponding values from a knowledge base.
        If a key is not found, it is replaced with a '?'.

        Args:
            s (str): The input string containing bracketed keys.
            knowledge (list[list[str]]): A list of key-value pairs, where each inner list is [key, value].

        Returns:
            str: The evaluated string with keys replaced by their values or '?'.
        """
        knowledgeMap = dict(knowledge)
        resultList = []
        currentKey = []
        isParsingKey = False

        for char in s:
            if char == '(':
                isParsingKey = True
            elif char == ')':
                isParsingKey = False
                keyString = "".join(currentKey)
                if keyString in knowledgeMap:
                    resultList.append(knowledgeMap[keyString])
                else:
                    resultList.append("?")
                currentKey = []
            elif isParsingKey:
                currentKey.append(char)
            else:
                resultList.append(char)

        return "".join(resultList)

File: Leetcode 1801-1900/1807/1807.java
// Leetcode 1807: Evaluate the Bracket Pairs of a String
// https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string/
// Solved on 27th of November, 2025
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    /**
     * Evaluates a string by replacing bracketed keys with their corresponding values from a knowledge base.
     * If a key is not found, it is replaced with a '?'.
     * @param s The input string containing bracketed keys.
     * @param knowledge A list of key-value pairs, where each inner list contains [key, value].
     * @return The evaluated string with keys replaced by their values or '?'.
     */
    public String evaluate(String s, List<List<String>> knowledge) {
        Map<String, String> map = new HashMap<>();
        for (List<String> pair : knowledge) {
            map.put(pair.get(0), pair.get(1));
        }
        
        StringBuilder sb = new StringBuilder();
        int n = s.length();
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (c == '(') {
                int j = s.indexOf(')', i + 1);
                String key = s.substring(i + 1, j);
                sb.append(map.getOrDefault(key, "?"));
                i = j;
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}